File: voice-orchestrator-windSurf-1.0/src/planner/taskPlanner.ts
import * as vscode from 'vscode';
import { Task, TaskStatus } from '../models/task';

export class TaskPlanner {
    private tasks: Map<string, Task> = new Map();
    private currentTaskId: string | null = null;
    private taskQueue: string[] = [];

    public async processCommand(command: string): Promise<void> {
        // Log the command
        console.log(`Processing command: ${command}`);
        
        // Create a new task
        const task = this.createTask(command);
        this.taskQueue.push(task.id);
        
        // If no task is currently running, start processing the queue
        if (!this.currentTaskId) {
            await this.processQueue();
        }
    }

    private createTask(description: string): Task {
        const taskId = `task-${Date.now()}`;
        const task: Task = {
            id: taskId,
            description,
            status: TaskStatus.Pending,
            createdAt: new Date(),
            updatedAt: new Date(),
            metadata: {}
        };
        
        this.tasks.set(taskId, task);
        return task;
    }

    private async processQueue(): Promise<void> {
        if (this.taskQueue.length === 0) {
            this.currentTaskId = null;
            return;
        }

        const taskId = this.taskQueue.shift()!;
        const task = this.tasks.get(taskId);
        
        if (!task) {
            // Skip if task not found
            return this.processQueue();
        }

        this.currentTaskId = taskId;
        task.status = TaskStatus.InProgress;
        task.updatedAt = new Date();
        
        try {
            // Show progress notification
            const progressOptions = {
                location: vscode.ProgressLocation.Notification,
                title: `Processing: ${task.description}`,
                cancellable: true
            };

            await vscode.window.withProgress(progressOptions, async (progress, token) => {
                token.onCancellationRequested(() => {
                    task.status = TaskStatus.Cancelled;
                    task.updatedAt = new Date();
                    vscode.window.showInformationMessage(`Task cancelled: ${task.description}`);
                });

                // Process the task
                await this.executeTask(task, progress, token);
            });

            task.status = TaskStatus.Completed;
            task.updatedAt = new Date();
            
        } catch (error) {
            console.error(`Error processing task ${task.id}:`, error);
            task.status = TaskStatus.Failed;
            task.error = error instanceof Error ? error.message : 'Unknown error';
            task.updatedAt = new Date();
            
            vscode.window.showErrorMessage(`Task failed: ${task.description}\n${task.error}`);
        } finally {
            // Process next task in the queue
            await this.processQueue();
        }
    }

    private async executeTask(task: Task, progress: vscode.Progress<{ message?: string; increment?: number }>, token: vscode.CancellationToken): Promise<void> {
        // Update progress
        progress.report({ message: 'Analyzing command...' });
        
        // Simulate work
        await new Promise(resolve => setTimeout(resolve, 1000));
        
        // Check for cancellation
        if (token.isCancellationRequested) {
            return;
        }
        
        // Here we would analyze the command and break it down into subtasks
        // For now, we'll just simulate this with a simple example
        const subtasks = this.analyzeCommand(task.description);
        
        // Process subtasks
        for (let i = 0; i < subtasks.length; i++) {
            const subtask = subtasks[i];
            progress.report({ 
                message: `Processing: ${subtask.description}`,
                increment: (1 / subtasks.length) * 100
            });
            
            // Simulate work for each subtask
            await new Promise(resolve => setTimeout(resolve, 1000));
            
            // Check for cancellation
            if (token.isCancellationRequested) {
                return;
            }
        }
        
        // Apply changes to the editor
        await this.applyChanges(task);
    }
    
    private analyzeCommand(command: string): Array<{ description: string; action: () => Promise<void> }> {
        // This is a simplified example - in a real implementation, this would use NLP
        // to understand the intent and break it down into actionable steps
        
        const lowerCommand = command.toLowerCase();
        const subtasks: Array<{ description: string; action: () => Promise<void> }> = [];
        
        if (lowerCommand.includes('login') && (lowerCommand.includes('page') || lowerCommand.includes('form'))) {
            subtasks.push(
                { 
                    description: 'Find login page files', 
                    action: async () => {
                        // Implementation would search for login-related files
                    } 
                },
                { 
                    description: 'Analyze current implementation', 
                    action: async () => {
                        // Implementation would analyze the current code
                    } 
                },
                { 
                    description: 'Update login form styling', 
                    action: async () => {
                        // Implementation would update CSS/JSX/HTML
                    } 
                }
            );
        } else if (lowerCommand.includes('api') && (lowerCommand.includes('create') || lowerCommand.includes('add'))) {
            subtasks.push(
                { 
                    description: 'Define API endpoint structure', 
                    action: async () => {
                        // Implementation would define API structure
                    } 
                },
                { 
                    description: 'Implement endpoint logic', 
                    action: async () => {
                        // Implementation would add the endpoint code
                    } 
                },
                { 
                    description: 'Add input validation', 
                    action: async () => {
                        // Implementation would add validation
                    } 
                },
                { 
                    description: 'Write tests', 
                    action: async () => {
                        // Implementation would add tests
                    } 
                }
            );
        } else {
            // Default subtask for unrecognized commands
            subtasks.push({
                description: `Process: ${command}`,
                action: async () => {
                    // Default action
                }
            });
        }
        
        return subtasks;
    }
    
    private async applyChanges(task: Task): Promise<void> {
        // This is where we would apply the actual changes to the code
        // For now, we'll just show a message with the proposed changes
        
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            vscode.window.showInformationMessage('No active editor');
            return;
        }
        
        // Example: Show a diff of the changes that would be made
        const document = editor.document;
        const position = editor.selection.active;
        
        // In a real implementation, we would use the AI to generate the code changes
        // For now, we'll just show a message with the proposed changes
        const changes = `Proposed changes for: ${task.description}\n\n` +
                       `- Update file: ${document.fileName}\n` +
                       `- At position: line ${position.line + 1}, character ${position.character + 1}\n` +
                       `- Action: ${task.description}`;
        
        // Show a preview of the changes
        const selection = await vscode.window.showInformationMessage(
            changes,
            'Apply Changes',
            'Preview Changes',
            'Cancel'
        );
        
        if (selection === 'Apply Changes') {
            // Apply the changes to the document
            await editor.edit(editBuilder => {
                // In a real implementation, this would apply the actual changes
                editBuilder.insert(position, `// ${task.description}\n`);
            });
            
            vscode.window.showInformationMessage('Changes applied successfully');
        } else if (selection === 'Preview Changes') {
            // Show a diff view of the changes
            // In a real implementation, this would show a proper diff view
            const diffDoc = await vscode.workspace.openTextDocument({
                content: `// Original content\n${document.getText()}\n\n` +
                         `// Proposed changes for: ${task.description}\n` +
                         `// [The actual changes would be shown here in a real implementation]`,
                language: document.languageId
            });
            
            await vscode.window.showTextDocument(diffDoc, { preview: false, viewColumn: vscode.ViewColumn.Beside });
        }
    }
    
    public getCurrentTask(): Task | null {
        if (!this.currentTaskId) {
            return null;
        }
        return this.tasks.get(this.currentTaskId) || null;
    }
    
    public getTaskQueue(): Task[] {
        return this.taskQueue.map(id => this.tasks.get(id)).filter(Boolean) as Task[];
    }
    
    public getCompletedTasks(): Task[] {
        return Array.from(this.tasks.values())
            .filter(task => task.status === TaskStatus.Completed)
            .sort((a, b) => (a.updatedAt > b.updatedAt ? -1 : 1));
    }
    
    public clearCompletedTasks(): void {
        const completedTaskIds = Array.from(this.tasks.entries())
            .filter(([_, task]) => task.status === TaskStatus.Completed)
            .map(([id]) => id);
            
        for (const id of completedTaskIds) {
            this.tasks.delete(id);
            const index = this.taskQueue.indexOf(id);
            if (index !== -1) {
                this.taskQueue.splice(index, 1);
            }
        }
    }
}

File: voice-orchestrator-windSurf-1.0/planner/task_planner.py
"""
Task Planning Module for Voice IDE Orchestrator.

This module handles breaking down high-level user commands into actionable tasks
that can be executed by the IDE's AI.
"""
from typing import List, Dict, Any, Optional
import logging
import json
from dataclasses import dataclass, asdict
from enum import Enum

logger = logging.getLogger(__name__)

class TaskStatus(Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"

@dataclass
class Task:
    """Represents a single task in the execution pipeline."""
    id: str
    description: str
    status: TaskStatus = TaskStatus.PENDING
    result: Optional[Any] = None
    error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert the task to a dictionary."""
        return {
            "id": self.id,
            "description": self.description,
            "status": self.status.value,
            "result": self.result,
            "error": self.error,
            "metadata": self.metadata or {}
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Task':
        """Create a Task from a dictionary."""
        return cls(
            id=data['id'],
            description=data['description'],
            status=TaskStatus(data.get('status', TaskStatus.PENDING.value)),
            result=data.get('result'),
            error=data.get('error'),
            metadata=data.get('metadata')
        )

class TaskPlanner:
    """Handles breaking down high-level commands into executable tasks."""
    
    def __init__(self):
        self.tasks: Dict[str, Task] = {}
        self.current_task_id: Optional[str] = None
    
    def create_task(self, description: str, metadata: Optional[Dict] = None) -> Task:
        """Create a new task with a unique ID."""
        task_id = f"task_{len(self.tasks) + 1}"
        task = Task(
            id=task_id,
            description=description,
            metadata=metadata or {}
        )
        self.tasks[task_id] = task
        return task
    
    def get_task(self, task_id: str) -> Optional[Task]:
        """Get a task by its ID."""
        return self.tasks.get(task_id)
    
    def update_task_status(self, task_id: str, status: TaskStatus, 
                         result: Any = None, error: Optional[str] = None) -> bool:
        """Update the status of a task."""
        if task_id not in self.tasks:
            logger.warning(f"Task {task_id} not found")
            return False
            
        task = self.tasks[task_id]
        task.status = status
        task.result = result
        task.error = error
        return True
    
    def plan_tasks(self, command: str) -> List[Task]:
        """Break down a high-level command into executable tasks.
        
        This is a simplified version that would be enhanced with AI in production.
        """
        command = command.lower()
        tasks = []
        
        # Simple keyword-based task planning
        if "login" in command and ("layout" in command or "style" in command):
            tasks = [
                self.create_task(
                    "Locate login page files",
                    {"type": "file_discovery", "patterns": ["*login*.{js,jsx,ts,tsx,html,css,scss}"]}
                ),
                self.create_task(
                    "Analyze current layout structure",
                    {"type": "code_analysis", "target": "login_component"}
                ),
                self.create_task(
                    "Update CSS for modern styling",
                    {"type": "code_modification", "target": "login_styles"}
                ),
                self.create_task(
                    "Ensure responsive design",
                    {"type": "code_modification", "target": "responsive_design"}
                ),
                self.create_task(
                    "Test across different screen sizes",
                    {"type": "testing", "target": "responsive_testing"}
                )
            ]
        elif "api" in command and ("add" in command or "create" in command):
            tasks = [
                self.create_task(
                    "Identify API endpoint requirements",
                    {"type": "analysis", "target": "api_requirements"}
                ),
                self.create_task(
                    "Create API endpoint implementation",
                    {"type": "code_creation", "target": "api_implementation"}
                ),
                self.create_task(
                    "Add input validation",
                    {"type": "code_modification", "target": "input_validation"}
                ),
                self.create_task(
                    "Write unit tests",
                    {"type": "testing", "target": "unit_tests"}
                )
            ]
        else:
            # Default task for unrecognized commands
            tasks = [
                self.create_task(
                    f"Process command: {command}",
                    {"type": "general", "command": command}
                )
            ]
        
        # Set the first task as current
        if tasks:
            self.current_task_id = tasks[0].id
            
        return tasks
    
    def get_current_task(self) -> Optional[Task]:
        """Get the currently active task."""
        if not self.current_task_id:
            return None
        return self.get_task(self.current_task_id)
    
    def complete_current_task(self, result: Any = None) -> Optional[Task]:
        """Mark the current task as completed and move to the next one."""
        if not self.current_task_id:
            return None
            
        self.update_task_status(
            self.current_task_id, 
            TaskStatus.COMPLETED,
            result=result
        )
        
        # Find the next pending task
        current_idx = -1
        task_list = list(self.tasks.values())
        
        for i, task in enumerate(task_list):
            if task.id == self.current_task_id:
                current_idx = i
                break
        
        next_task = None
        for task in task_list[current_idx + 1:]:
            if task.status == TaskStatus.PENDING:
                next_task = task
                break
        
        self.current_task_id = next_task.id if next_task else None
        return next_task
    
    def get_status_summary(self) -> Dict[str, Any]:
        """Get a summary of all tasks and their statuses."""
        return {
            "total_tasks": len(self.tasks),
            "completed": sum(1 for t in self.tasks.values() if t.status == TaskStatus.COMPLETED),
            "in_progress": sum(1 for t in self.tasks.values() if t.status == TaskStatus.IN_PROGRESS),
            "pending": sum(1 for t in self.tasks.values() if t.status == TaskStatus.PENDING),
            "current_task": self.current_task_id,
            "tasks": [t.to_dict() for t in self.tasks.values()]
        }
    
    def clear_completed_tasks(self) -> None:
        """Remove completed tasks to free up memory."""
        completed_ids = [
            task_id for task_id, task in self.tasks.items()
            if task.status == TaskStatus.COMPLETED
        ]
        
        for task_id in completed_ids:
            if self.current_task_id == task_id:
                self.current_task_id = None
            del self.tasks[task_id]
        
        logger.info(f"Cleared {len(completed_ids)} completed tasks")


# Example usage
if __name__ == "__main__":
    import logging
    logging.basicConfig(level=logging.INFO)
    
    planner = TaskPlanner()
    
    # Example command processing
    command = "Fix the layout issue on the login page and make it look more modern"
    print(f"Processing command: {command}")
    
    tasks = planner.plan_tasks(command)
    print(f"\nPlanned {len(tasks)} tasks:")
    for task in tasks:
        print(f"- {task.description}")
    
    # Simulate task execution
    print("\nSimulating task execution...")
    for task in tasks:
        print(f"\nExecuting: {task.description}")
        planner.update_task_status(task.id, TaskStatus.IN_PROGRESS)
        # Simulate work
        import time
        time.sleep(0.5)
        planner.complete_current_task("Task completed successfully")
    
    # Print final status
    print("\nFinal status:")
    print(json.dumps(planner.get_status_summary(), indent=2))
